### PR TITLE
GCP DNS Data Connector

### PR DESCRIPTION
**Pull request by vu-socprime**
## Proposed Changes

  - add GCP DNS Data Connector

### CODE REVIEW

File: Solutions/GoogleCloudPlatformDNS/Data Connectors/GCP_DNS_API_FunctionApp.json [hunk 1]
```diff
@@ -0,0 +1,130 @@
+{
+    "id": "GCPDNSDataConnector",
+    "title": "Google Cloud Platform DNS",
+    "publisher": "Google",
+    "descriptionMarkdown": "The Google Cloud Platform DNS data connector provides the capability to ingest [Cloud DNS query logs](https://cloud.google.com/dns/docs/monitoring#using_logging) [and Cloud DNS audit logs](https://cloud.google.com/dns/docs/audit-logging) into Azure Sentinel using the GCP Logging API. Refer to [GCP Logging API documentation](https://cloud.google.com/logging/docs/api) for more information.",
+    "additionalRequirementBanner": ">This data connector depends on a parser based on a Kusto Function to work as expected [**imDNSGCPCloudDNS**](https://aka.ms/sentinel-GCPDNSDataConnector-parser) which is deployed with the Azure Sentinel Solution.",
```
Comment (v-admahe) [inline Solutions/GoogleCloudPlatformDNS/Data Connectors/GCP_DNS_API_FunctionApp.json:6]: As per the latest template there is change in additional requirement banner text could yo please update it. You can take a reference from [here](https://github.com/Azure/Azure-Sentinel/blob/master/DataConnectors/Templates/Connector_REST_API_AzureFunctionApp_template/DataConnector_API_AzureFunctionApp_template.json)

Thanks,
Aditi Maheshwari

File: Solutions/GoogleCloudPlatformDNS/Data Connectors/GCP_DNS_API_FunctionApp.json [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            "title": "",
+            "description": ">**NOTE:** This connector uses Azure Functions to connect to the GCP API to pull logs into Azure Sentinel. This might result in additional data ingestion costs. Check the [Azure Functions pricing page](https://azure.microsoft.com/pricing/details/functions/) for details."
+        },
+        {
+            "title": "",
+            "description": ">**(Optional Step)** Securely store workspace and API authorization key(s) or token(s) in Azure Key Vault. Azure Key Vault provides a secure mechanism to store and retrieve key values. [Follow these instructions](https://docs.microsoft.com/azure/app-service/app-service-key-vault-references) to use Azure Key Vault with an Azure Function App."
+        },
+        {
+            "title": "",
+            "description": ">**NOTE:** This data connector depends on a parser based on a Kusto Function to work as expected [**imDNSGCPCloudDNS**](https://aka.ms/sentinel-GCPDNSDataConnector-parser) which is deployed with the Azure Sentinel Solution."
```
Comment (v-admahe) [inline Solutions/GoogleCloudPlatformDNS/Data Connectors/GCP_DNS_API_FunctionApp.json:83]: As per the latest template there is change in additional requirement banner text so we have to update here as well could you please update it. You can take a reference from [here](https://github.com/Azure/Azure-Sentinel/blob/master/DataConnectors/Templates/Connector_REST_API_AzureFunctionApp_template/DataConnector_API_AzureFunctionApp_template.json)

Thanks,
Aditi Maheshwari

**Review by v-admahe — CHANGES_REQUESTED** [review]

File: Sample Data/Custom/GCP_DNS_CL.json [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+        "payload_authorizationInfo_s": "",
+        "payload_resourceName_s": "",
+        "payload_request__type_s": "",
+        "payload_request_project_s": "",
+        "resource_labels_policy_name_s": "",
+        "payload_request_managedZone_s": "",
+        "log_name_s": "projects/sometestproject-1111111/logs/dns.googleapis.com%2Fdns_queries",
+        "insert_id_s": "785cqudte9v",
+        "severity_s": "INFO",
+        "timestamp_t [UTC]": "6/17/2021, 9:15:33.277 AM",
```
Comment (v-admahe) [inline Sample Data/Custom/GCP_DNS_CL.json:25]: "timestamp_t [UTC]" fields value ingesting in LAW as string , due to which parser is failing, can you please provide both fields values as per following format in all places.
"timestamp_t [UTC]":"2021-06-16T02:27:00.00000Z",
Also please remove the [UTC] from the column name.

Thanks,
Aditi Maheshwari

**Review by v-admahe — CHANGES_REQUESTED** [review]

File: Sample Data/Custom/GCP_DNS_CL.json (hunk 3 above)
Comment (vu-socprime) [reply Sample Data/Custom/GCP_DNS_CL.json]: updated

**Comment by oshezaf** [comment]
Hi @v-admahe , @vu-socprime : 

I took the liberty of enhancing the parser to be fully compatible with the new DNS schema. I did not want to go and override the git page, so including it here. Also, not having a workspace with the data, I did not test it, so may have typos.

      let DNSQuery_GcpDns=(){
          GCP_DNS_CL
          | project-rename
             Query=payload_queryName_s, 
             QueryTypeName=payload_queryType_s,
             ResponseName=payload_rdata_s, 
             EventResultDetails=payload_responseCode_s,
             NetworkProtocol=payload_protocol_s, 
             SrcIpAddr=payload_sourceIP_s,
             EventOriginalUid=insert_id_s,
             EventSeverity=severity_s
         
          | extend
             EventCount=int(1),
             EventProduct='Cloud DNS',
             EventVendor='GCP',
             EventSchemaVersion="0.1.0",
             Dvc="GCPDNS" ,
             EventType = iif (resource_type_s == "dns_query", "lookup", resource_type_s)
             EventResult=iff(payload_responseCode_s=~'NOERROR','Success','Failure'),
             EventSubType='response',
             EventEndTime=todatetime(Timestamp_t)
           
          // -- Aliases
          | extend 
             ResponseCodeName=EventResultDetails, 
             Domain=Query,
             IpAddr=SrcIpAddr
             EventStartTime = EventEndTime
       
             };
      DNSQuery_GcpDns

**Comment by v-jayakal** [comment]
@vu-socprime , can you please update the parser by checking the function text provided by @v-maudan, thanks.

**Comment by vu-socprime** [comment]
> @vu-socprime , can you please update the parser by checking the function text provided by @v-maudan, thanks.

@v-jayakal Parser updated. Thanks.

**Comment by v-maudan** [comment]
@vu-socprime , It seems parser not used anywhere, Could you please confirm on this ?

**Comment by vu-socprime** [comment]
> @vu-socprime ,
> 
> * It seems parser not used anywhere, Could you please confirm on this ?
> * `Workbooks/Images/Preview/UniFiSGWhite.png` Please remove this file from PR.

@v-maudan 1. Yes, parser is not used in current PR.
2. Workbooks/Images/Preview/UniFiSGWhite.png removed

**Review by v-maudan — DISMISSED** [review]

**Comment by v-maudan** [comment]
> . Workbooks/Images/Preview/UniFiSGWhite.png removed

@vu-socprime , Still I am able to see workbook image in this PR, can you please check and remove. Thanks!

**Review by v-jayakal — APPROVED** [review]

**Review by v-jayakal — APPROVED** [review]

**Pull request merged**